python/imzml_reader.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+
+import numpy as np
+import pandas as pd
+from pyimzml.ImzMLParser import ImzMLParser
+import zarr
+
+import argparse
+
+
+class IMSDataset:
+    """Converts IMS data stored as imzML into columnar or ndarray formats
+
+    :param imzml_file: path to `.imzML` file.
+    :param ibd_file: path to associated `.ibd` file.
+    :param micro_res: microscopy resolution in nm (used for scaling).
+    :param ims_res: IMS resolution in nm (used for scaling).
+    """
+
+    def __init__(self, imzml_file, ibd_file, micro_res, ims_res):
+        # When passing the ibd path explicitly,
+        # the file object must be opened manually
+        self.parser = ImzMLParser(
+            filename=imzml_file, ibd_file=open(ibd_file, "rb")
+        )
+        self.micro_res = micro_res
+        self.ims_res = ims_res
+        self.ims_px_in_micro = ims_res / micro_res  # scaling factor
+
+        mz_lengths = self.parser.mzLengths
+        if not (mz_lengths.count(mz_lengths[0]) == len(mz_lengths)):
+            raise ValueError(
+                "The number of m/z is not the same at each coordinate."
+            )
+        self.mzs, _ = self.parser.getspectrum(0)
+
+    def __get_min_max_coords(self):
+        coords = np.array(self.parser.coordinates)
+        x_min, y_min, _ = np.min(coords, axis=0)
+        x_max, y_max, _ = np.max(coords, axis=0)
+        return x_min, y_min, x_max, y_max
+
+    def __format_mzs(self, precision=4):
+        return np.round(self.mzs, precision).astype(str)
+
+    def to_columnar(self, dtype="uint32"):
+        coords = np.array(self.parser.coordinates)
+        x, y, _ = coords.T
+
+        coords_df = pd.DataFrame(
+            {
+                "x": x,
+                "y": y,
+                "micro_x_topleft": x * self.ims_px_in_micro
+                - self.ims_px_in_micro,
+                "micro_y_topleft": y * self.ims_px_in_micro
+                - self.ims_px_in_micro,
+                "micro_px_width": np.repeat(self.ims_px_in_micro, len(coords)),
+            },
+            dtype=dtype,
+        )
+
+        intensities = np.zeros((len(coords_df), len(self.mzs)))
+        for i in range(len(coords)):
+            _, coord_intensities = self.parser.getspectrum(i)
+            intensities[i, :] = coord_intensities
+
+        intensities = pd.DataFrame(
```

Comment:
Prefer a new variable name here. If this gets longer over time, it might be easy to overlook that the `intensities` at the end is not the same kind of thing as in the initial assignment.

Reply:
Agreed.